### PR TITLE
fully_shard load state_dict

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90945

Ensures that load_state_dict for fully_shard works:
- Don't add back FSDP prefix
- Small fix to ensure mixed precision check for buffers work

Follow ups:
- state_dict_type does not work, blocking rank0_only and CPU offload as well as other state dict implementations
- No testing when wrapped with AC, using mixed precision, integration with distributed checkpoint, etc.

### CODE REVIEW

File: torch/testing/_internal/common_fsdp.py [hunk 1]
```diff
@@ -137,6 +137,21 @@ def subtest_name(test_name_mapping, *args):
         [test_name_mapping[str(s)] if s is not None else "none" for s in args]
     )
 
+def _broadcast_state_dict(rank, state_dict):
```
Comment (rohan-varma) [inline torch/testing/_internal/common_fsdp.py:140]: not strictly needed right now but will be used in composable tests.

File: test/distributed/_composable/test_fully_shard.py [hunk 2]
```diff
@@ -452,44 +455,148 @@ def world_size(self) -> int:
...
+        load_composable = fully_shard(
+            copy.deepcopy(local_model),
+            policy=ModuleWrapPolicy({UnitModule})
+        )
+        _zero_model(load_composable)
+        for p in load_composable.parameters():
+            self.assertEqual(p.sum(), 0)
+
+        sd = {k: v.clone() for k, v in composable_sd.items()}
+        print(f"sd keys: {list(sd.keys())}")
```
Comment (awgu) [inline test/distributed/_composable/test_fully_shard.py:482]: nit: Remember to remove.

File: test/distributed/_composable/test_fully_shard.py [hunk 3]
```diff
@@ -452,44 +455,148 @@ def world_size(self) -> int:
...
+            self.assertEqual(0, p.sum())
+
+        sd = {k: v.clone() for k, v in composable_sd.items()}
+        load_composable.load_state_dict(sd)
+        self._check_model_parity(load_composable, save_composable)
+
+    @skip_if_lt_x_gpu(2)
+    def test_state_dict_save_load_flow(self):
+        """
+        E2E test of save + load with rank0_only + CPU offload for TransformerWithSharedParams
```
Comment (awgu) [inline test/distributed/_composable/test_fully_shard.py:521]: In the future, will this test include different state dict types and subtest the different configs?

File: torch/distributed/fsdp/_state_dict_utils.py [hunk 4]
```diff
@@ -237,14 +237,20 @@ def _common_unshard_post_state_dict_hook(
             # TODO: for composable FSDP, this should be clean_tensor_name(clean_key),
             buffer_clean_fqns.append(clean_key)
             buffers.append(state_dict[fqn])
-    if buffers and fsdp_state._mixed_precision_enabled_for_buffers():
-        buffer_dtypes = _get_buffer_dtypes(fsdp_state, buffer_clean_fqns)
-        _cast_buffers_to_dtype_and_device(
-            buffers, buffer_dtypes, fsdp_state.compute_device
+    if buffers:
+        mixed_precision_enabled_for_buffers = (
+            fsdp_state._mixed_precision_enabled_for_buffers() if not _is_composable(fsdp_state)
```
Comment (awgu) [inline torch/distributed/fsdp/_state_dict_utils.py:242]: To-do: We can make `_mixed_precision_enabled_for_buffers()` not be a method of `FullyShardedDataParallel` to make this not have to `if` / `else` here. We would be able to just check `fsdp_state.mxied_precision.buffer_dtype is not None` -- flexible whether that is in its own function or written like that every time.

**Review by awgu — APPROVED** [review]
LGTM! There a few to-dos you left for yourself. Feel free to address those before landing.

File: torch/testing/_internal/common_fsdp.py [hunk 5]
```diff
@@ -111,15 +111,18 @@ def _assert_module_states(
...
     zero_buffers: bool = False,
+    summon_full=True,
 ):
     """Zeros the parameters and optionally buffers of ``model`` in place."""
-    for param in model.parameters():
-        with torch.no_grad():
-            param.zero_()
-    if zero_buffers:
-        for buffer in model.buffers():
+    ctx = FSDP.summon_full_params(model) if summon_full else suppress()
```
Comment (awgu) [inline torch/testing/_internal/common_fsdp.py:117]: Should we include any to-do or issue for following up on this? Or, could you remind me what the current status on this is?

**Comment by rohan-varma** [comment]
lint jobs are unrelated and failing due to clang-tidy binary hash issue.

File: torch/testing/_internal/common_fsdp.py (hunk 5 above)
Comment (rohan-varma) [reply torch/testing/_internal/common_fsdp.py:117]: yeah, let me file an issue summarizing it.

**Pull request closed without merging**